sfge/src/main/java/com/salesforce/graph/ops/expander/PathExpansionRegistry.java

```diff
@@ -0,0 +1,119 @@
+package com.salesforce.graph.ops.expander;
+
+import com.google.common.collect.ImmutableMap;
+import com.salesforce.graph.ApexPath;
+import com.salesforce.graph.ops.registry.AbstractRegistryData;
+import com.salesforce.graph.ops.registry.Indexable;
+import com.salesforce.graph.ops.registry.Registry;
+import java.util.Map;
+import java.util.function.Supplier;
+import org.apache.tinkerpop.gremlin.process.traversal.dsl.graph.GraphTraversalSource;
+
+/**
+ * Optimization to maintain lower heap space usage while expanding paths. See {@link Registry} for
+ * information on how the optimization works.
+ *
+ * <p>This registry implementation holds data for {@link ApexPathCollapser}, {@link
+ * ApexPathExpander}, and {@link ForkEvent} instances. Collections that need to hold instances of
+ * these types, can hold their {@link Long} Ids instead, and use the corresponding <code>lookup()
+ * </code> method to fetch the instance from registry.
+ *
+ * <p>{@link PathExpansionRegistry} uses instances of {@link AbstractRegistryData} rather than
+ * {@link ThreadLocal} so that these hashmaps only grow as needed per path expansion of an ApexPath.
+ * Once cleared, all heap usage will be released back. This makes sure:
+ * <ul>
+ *     <li>1. Registry data lives for a shorted duration of time and doesn't move to Old Gen of heap space until later. This allows
+ *  earlier clean up from heap space when they are not used.</li>
+ *     <li>2. Internally-used HashMaps can start
+ *  with smaller sizes and can grow as needed, as opposed to starting with a much larger HashMap that
+ *  was needed in a previous path expansion.</li>
+ * </ul>
+ *
+ *
+ * <p>Create a new instance of this registry when a path expansion for an ApexPath is initiated at
+ * {@link ApexPathExpanderUtil#expand(GraphTraversalSource, ApexPath, ApexPathExpanderConfig)}.
+ * Clear the registry when the operation completes to avoid accidental memory leaks.
+ */
+public class PathExpansionRegistry extends Registry {
+
+    private static final Map<Class<? extends Indexable>, Supplier> REGISTRY_SUPPLIER =
```

Comment:
It feels odd to me to have a registry cover multiple classes in this way. If we needed to implement more registries, would we do that in another class? I feel like that could lead to confusion over which classes are registered where.
Should we maybe try to make the top-level `Registry` class generic enough that it could act as a registry for any implementation of `Indexable`? Alternatively, perhaps refactor it so each Registry is associated with only one class?
Do either of those make sense?

Reply:
I see what you mean. The reason I've broken down the layers is to give us more flexibility to define the timeline of a Registry instance's existence.

For example, in the case of path expansion, we want a registry to hold data only for one ApexPath per entry point. The registry concept here would be counter-productive if it remains alive the entire time Graph Engine executes since we want the heap space to be maintained and released as soon as possible.
There could be another area in the near future that's probably related to executing rules. At that time, we may want to use the registry for a different timeline when a particular rule finds an interesting vertex. Or another case where we use it as a singleton across all the threads available for the entire execution time.

Having the framework separate from the instantiation gives us options to use it according to our specific requirements.

Reply:
That makes sense.